File: client/kotlin/nona-client/src/main/kotlin/com/nonaconfig/client/NonaConfig.kt
package com.nonaconfig.client

import android.content.Context
import java.io.File
import java.io.IOException
import java.net.URLEncoder
import java.util.concurrent.CompletableFuture
import kotlin.time.Duration
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.coroutines.future.future
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import org.json.JSONObject

/**
 * Reads Nona config on Android. Values are held in memory and read
 * synchronously, downloading is separate from applying, and the last good
 * snapshot survives restarts.
 *
 * Every value is the same for every user; Nona has no runtime targeting.
 */
class NonaConfig internal constructor(
    private val options: NonaOptions,
    private val http: NonaHttpClient,
    private val store: NonaSnapshotStore,
    private val clock: () -> Long,
    private val ioDispatcher: CoroutineDispatcher,
) {

    private val fetchLock = Mutex()
    private val stateLock = Any()
    // Cache operations serialize before state mutation; never perform I/O under stateLock.
    private val cacheLock = Mutex()
    private var generation = 0L
    private val cacheIdentity = options.cacheIdentity()

    @Volatile
    private var active: Snapshot? = null

    @Volatile
    private var pending: Snapshot? = null

    @Volatile
    private var defaults: Map<String, NonaEntry> = emptyMap()

    @Volatile
    private var lastFetchAtMillis: Long = 0

    private val updates = ConfigUpdates()

    /** Per-collector activation invalidations. Slow collectors receive a union of changed keys. */
    val configUpdates: Flow<Set<String>> = updates.flow

    /** Keys readable right now, from the active snapshot and the defaults. */
    val keys: Set<String>
        get() = LinkedHashSet<String>().apply {
            active?.values?.keys?.let(::addAll)
            addAll(defaults.keys)
        }

    /** Values used until a key arrives from Nona. Accepts booleans, numbers and strings. */
    fun setDefaults(values: Map<String, Any>) {
        defaults = values.entries.associate { (key, value) -> key to value.toEntry() }
    }

    /**
     * Loads the persisted snapshot into the active set; no network work. Call
     * once at startup, before the first read. Returns true if one was restored.
     */
    suspend fun initialize(): Boolean = withContext(ioDispatcher) {
        cacheLock.withLock {
            if (synchronized(stateLock) { active != null || pending != null }) return@withLock false
            val restored = store.read()?.let { Snapshot.fromCacheJson(it, cacheIdentity) }
                ?: return@withLock false
            val context = currentCoroutineContext()
            synchronized(stateLock) {
                context.ensureActive()
                active = restored
                lastFetchAtMillis = restored.fetchedAtMillis
            }
            true
        }
    }

    /**
     * Downloads into the pending slot; reads keep the activated values until
     * [activate].
     *
     * @throws NonaHttpException if Nona rejects the request.
     * @throws NonaException on a transport or parsing failure.
     */
    suspend fun fetch(minimumFetchInterval: Duration = options.minimumFetchInterval): FetchStatus =
        fetchLock.withLock {
            require(minimumFetchInterval.isFinite() && minimumFetchInterval >= Duration.ZERO) {
                "minimumFetchInterval must be finite and nonnegative"
            }
            val now = clock()
            if (lastFetchAtMillis > 0 && now >= lastFetchAtMillis &&
                now - lastFetchAtMillis < minimumFetchInterval.inWholeMilliseconds
            ) {
                return@withLock FetchStatus.THROTTLED
            }

            withContext(ioDispatcher) { load(now) }
        }

    /** Promotes fetched values into the active set. Returns true if they changed. */
    fun activate(): Boolean {
        val changed = synchronized(stateLock) {
            val next = pending ?: return false
            pending = null
            changedKeys(active?.values.orEmpty(), next.values).also { active = next }
        }
        // Collectors can run inline; never invoke them while holding stateLock.
        if (changed.isNotEmpty()) updates.publish(changed)
        return changed.isNotEmpty()
    }

    /** [fetch] followed by [activate]. Returns true if the active values changed. */
    suspend fun fetchAndActivate(
        minimumFetchInterval: Duration = options.minimumFetchInterval,
    ): Boolean {
        if (fetch(minimumFetchInterval) == FetchStatus.DISCARDED) return false
        return activate()
    }

    /** Forgets the cached snapshot on disk and in memory. */
    suspend fun reset() = withContext(ioDispatcher) {
        cacheLock.withLock {
            val context = currentCoroutineContext()
            synchronized(stateLock) {
                context.ensureActive()
                generation++
                active = null
                pending = null
                lastFetchAtMillis = 0
            }
            store.clear()
        }
    }

    /** Java-friendly asynchronous counterparts; cancel the future to cancel the coroutine. */
    fun initializeAsync(): CompletableFuture<Boolean> = async { initialize() }
    fun fetchAsync(): CompletableFuture<FetchStatus> = async { fetch() }
    fun fetchAndActivateAsync(): CompletableFuture<Boolean> = async { fetchAndActivate() }
    fun resetAsync(): CompletableFuture<Void?> = async { reset(); null }

    fun getBoolean(key: String): Boolean = valueOrDefault(key, ValueParsing::parseBoolean, false)

    fun getString(key: String): String = valueOrDefault(key, ValueParsing::parseString, "")

    fun getLong(key: String): Long = valueOrDefault(key, ValueParsing::parseLong, 0L)

    fun getDouble(key: String): Double = valueOrDefault(key, ValueParsing::parseDouble, 0.0)

    private fun <T> async(block: suspend () -> T): CompletableFuture<T> =
        CoroutineScope(ioDispatcher).future { block() }

    private fun <T> valueOrDefault(
        key: String,
        parse: (String, String) -> ValueParsing.ParseResult<T>,
        zero: T,
    ): T {
        val remote = active?.values?.get(key)?.let { parse(key, it.value) }
        if (remote is ValueParsing.ParseResult.Ok) return remote.value
        val fallback = defaults[key]?.let { parse(key, it.value) }
        return if (fallback is ValueParsing.ParseResult.Ok) fallback.value else zero
    }

    /** Raw entry origin; a typed getter may fall back if that entry cannot be parsed. */
    fun getSource(key: String): NonaValueSource =
        entryFor(key)?.second ?: NonaValueSource.STATIC

    fun resolveBoolean(key: String): NonaResolution<Boolean> =
        resolve(key, ValueParsing::parseBoolean)

    /**
     * Resolves the raw stored string. JSON values come back unparsed, so an
     * OpenFeature provider can decode them into whatever shape it needs.
     */
    fun resolveString(key: String): NonaResolution<String> =
        resolve(key, ValueParsing::parseString)

    fun resolveLong(key: String): NonaResolution<Long> =
        resolve(key, ValueParsing::parseLong)

    fun resolveDouble(key: String): NonaResolution<Double> =
        resolve(key, ValueParsing::parseDouble)

    private fun <T> resolve(
        key: String,
        parse: (String, String) -> ValueParsing.ParseResult<T>,
    ): NonaResolution<T> {
        val (entry, source) = entryFor(key) ?: return missing(key)

        return when (val parsed = parse(key, entry.value)) {
            is ValueParsing.ParseResult.Ok -> NonaResolution.Success(
                value = parsed.value,
                source = source,
                contentType = entry.contentType,
            )

            is ValueParsing.ParseResult.Err ->
                NonaResolution.Failure(parsed.reason, parsed.message)
        }
    }

    /** The single place that knows a remote value outranks a default. */
    private fun entryFor(key: String): Pair<NonaEntry, NonaValueSource>? =
        active?.values?.get(key)?.let { it to NonaValueSource.REMOTE }
            ?: defaults[key]?.let { it to NonaValueSource.DEFAULT }

    private fun missing(key: String): NonaResolution.Failure = if (active == null) {
        NonaResolution.Failure(
            NonaFailureReason.NOT_READY,
            "No Nona snapshot has been loaded yet, and '$key' has no default.",
        )
    } else {
        NonaResolution.Failure(
            NonaFailureReason.NOT_FOUND,
            "Nona key '$key' is not in the '${options.environmentId}' snapshot. " +
                "Client-side reads only see frontend-scoped entries.",
        )
    }

    private suspend fun load(now: Long): FetchStatus {
        val (requestGeneration, previous) = synchronized(stateLock) { generation to (pending ?: active) }
        val headers = buildMap {
            options.apiKey?.let { put("X-Api-Key", it) }
            previous?.etag?.let { put("If-None-Match", it) }
        }

        val response = try {
            http.get(snapshotUrl(), headers)
        } catch (cause: IOException) {
            throw NonaException("Nona request failed.", cause)
        }
        currentCoroutineContext().ensureActive()
        when (response.statusCode) {
            in 200..299 -> Unit

            HTTP_NOT_MODIFIED -> {
                if (previous == null) throw NonaException("Nona returned 304 without a cached snapshot.")
                return persist(previous.copy(fetchedAtMillis = now), requestGeneration, revalidated = true)
            }

            else -> {
                val problem = parseProblemDetails(response.body)
                throw NonaHttpException(
                    statusCode = response.statusCode,
                    message = problem.detail ?: describeHttpFailure(response.statusCode),
                    errorCode = problem.errorCode,
                    detail = problem.detail,
                )
            }
        }

        if (response.body.toByteArray(Charsets.UTF_8).size > options.maxResponseBytes) {
            throw NonaException("Nona snapshot exceeds maxResponseBytes (${options.maxResponseBytes}).")
        }
        val snapshot = Snapshot.fromResponseBody(response.body, response.etag, now)
        return persist(snapshot, requestGeneration, revalidated = false)
    }

    private suspend fun persist(snapshot: Snapshot, requestGeneration: Long, revalidated: Boolean): FetchStatus =
        cacheLock.withLock {
            val json = snapshot.toCacheJson(cacheIdentity)
            val context = currentCoroutineContext()
            synchronized(stateLock) {
                context.ensureActive()
                if (generation != requestGeneration) return@withLock FetchStatus.DISCARDED
                if (revalidated && pending == null) active = snapshot else pending = snapshot
                lastFetchAtMillis = snapshot.fetchedAtMillis
            }
            store.write(json)
            if (revalidated) FetchStatus.NOT_MODIFIED else FetchStatus.SUCCESS
        }

    private fun describeHttpFailure(statusCode: Int): String = when (statusCode) {
        401 -> "Nona rejected the API key."
        404 -> "Nona has no snapshot for environment '${options.environmentId}'. " +
            "Check the environment id, and that the API key is frontend-scoped: " +
            "the snapshot endpoint is hidden from backend-only keys."

        else -> "Nona returned HTTP $statusCode."
    }

    private fun snapshotUrl(): String {
        val base = options.normalizedBaseUrl()
        val releaseVersion = options.releaseVersion?.trim()?.takeIf { options.useReleases && it.isNotEmpty() }
        val query = buildList {
            options.prefix?.let { add("prefix=" + encode(it)) }
        }.joinToString("&")

        val path = if (options.useReleases) {
            val release = releaseVersion?.let(::encode) ?: "active"
            "$base/api/environments/${encode(options.environmentId)}/releases/$release/parameters"
        } else {
            "$base/api/environments/${encode(options.environmentId)}/parameters"
        }
        return if (query.isEmpty()) path else "$path?$query"
    }

    private fun parseProblemDetails(body: String): ProblemDetails {
        if (body.isBlank()) return ProblemDetails()
        return try {
            val json = JSONObject(body)
            ProblemDetails(
                errorCode = json.opt("errorCode")?.takeIf { it is String } as? String,
                detail = json.opt("detail")?.takeIf { it is String } as? String,
            )
        } catch (_: Exception) {
            ProblemDetails()
        }
    }

    companion object {
        private const val HTTP_NOT_MODIFIED = 304

        /** Creates an instance that caches its snapshot in the app's private files. */
        @JvmStatic
        fun create(context: Context, options: NonaOptions): NonaConfig =
            create(
                options = options,
                store = FileSnapshotStore(
                    File(context.filesDir, "nona/${cacheFileName(options)}"),
                ),
            )

        /** Creates an instance with a supplied store and, optionally, HTTP client. */
        @JvmStatic
        @JvmOverloads
        fun create(
            options: NonaOptions,
            store: NonaSnapshotStore,
            http: NonaHttpClient = UrlConnectionHttpClient(
                connectTimeoutMillis = options.connectTimeout.inWholeMilliseconds.toInt(),
                readTimeoutMillis = options.readTimeout.inWholeMilliseconds.toInt(),
                maxResponseBytes = options.maxResponseBytes,
            ),
            clock: () -> Long = System::currentTimeMillis,
            ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
        ): NonaConfig = NonaConfig(options, http, store, clock, ioDispatcher)

        /** Bound to origin, key/project and snapshot selectors, without exposing the key. */
        private fun cacheFileName(options: NonaOptions): String {
            return "snapshot-${options.cacheIdentity()}.json"
        }
    }
}

private data class ProblemDetails(
    val errorCode: String? = null,
    val detail: String? = null,
)

private fun encode(value: String): String =
    URLEncoder.encode(value, "UTF-8").replace("+", "%20")

private fun Any.toEntry(): NonaEntry = when (this) {
    is Boolean -> NonaEntry(toString(), "boolean")
    is Number -> NonaEntry(toString(), "number")
    is String -> NonaEntry(this, "text")
    else -> NonaEntry(toString(), "text")
}

File: client/swift/Sources/NonaClient/NonaConfig.swift
import Foundation

/// Thread-safe synchronous reads with explicit asynchronous fetch and activation.
public final class NonaConfig: Sendable {
    private struct State {
        var active: Snapshot?
        var pending: Snapshot?
        var defaults: [String: NonaEntry] = [:]
        var generation: UInt64 = 0
        var lastFetch: TimeInterval?
        var subscribers: [UUID: UpdateSubscriber] = [:]
    }

    public let options: NonaOptions
    private let http: any NonaHTTPClient
    private let store: any NonaSnapshotStore
    private let clock: @Sendable () -> TimeInterval
    private let state = Locked(State())
    // Lock ordering is cacheAccess -> state. Synchronous readers never acquire cacheAccess.
    private let cacheAccess = SerialAccess()
    private let fetchGate = FetchGate()
    private let identity: String

    public init(options: NonaOptions, store: (any NonaSnapshotStore)? = nil,
                http: (any NonaHTTPClient)? = nil,
                clock: @escaping @Sendable () -> TimeInterval = { Date().timeIntervalSince1970 }) {
        self.options = options
        self.identity = options.cacheIdentity
        self.http = http ?? URLSessionHTTPClient(options: options)
        self.clock = clock
        if let store {
            self.store = store
        } else if let directory = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first {
            self.store = FileSnapshotStore(fileURL: directory.appendingPathComponent("NonaClient", isDirectory: true)
                .appendingPathComponent("snapshot-\(identity).json"))
        } else {
            self.store = InMemorySnapshotStore()
        }
    }

    deinit {
        let subscribers = state.withLock { Array($0.subscribers.values) }
        subscribers.forEach { $0.finish() }
    }

    public var keys: Set<String> {
        state.withLock { Set(($0.active?.values ?? [:]).keys).union($0.defaults.keys) }
    }

    public func setDefaults(_ values: [String: NonaDefaultValue]) {
        state.withLock { $0.defaults = values.mapValues(\.entry) }
    }

    /// A separate stream per subscriber. Slow consumers receive the union of buffered changed keys.
    public func updates() -> AsyncStream<Set<String>> {
        let id = UUID()
        return AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            state.withLock { $0.subscribers[id] = UpdateSubscriber(continuation) }
            continuation.onTermination = { [weak self] _ in
                self?.state.withLock { _ = $0.subscribers.removeValue(forKey: id) }
            }
        }
    }

    /// Restores the last successful fetch, without a network request. Call before the first fetch.
    @discardableResult
    public func initialize() async throws -> Bool {
        try await background { [self] in
            try cacheAccess.withLock {
                try Task.checkCancellation()
                guard state.withLock({ $0.active == nil && $0.pending == nil }),
                      let data = store.read(),
                      let snapshot = try? JSONDecoder().decode(Snapshot.self, from: data),
                      snapshot.identity == identity, snapshot.fetchedAt.isFinite else { return false }
                return try state.withLock { state in
                    try Task.checkCancellation()
                    state.active = snapshot
                    state.lastFetch = snapshot.fetchedAt
                    return true
                }
            }
        }
    }

    @discardableResult
    public func fetch(minimumFetchInterval: TimeInterval? = nil) async throws -> NonaFetchStatus {
        let interval = minimumFetchInterval ?? options.minimumFetchInterval
        guard interval.isFinite, interval >= 0 else {
            throw NonaError.invalidOptions("minimumFetchInterval must be finite and nonnegative.")
        }
        try await fetchGate.acquire()
        do {
            try Task.checkCancellation()
            let result = try await load(minimumFetchInterval: interval)
            await fetchGate.release()
            return result
        } catch {
            await fetchGate.release()
            throw error
        }
    }

    /// Returns true when an activated value or its content type changed.
    @discardableResult
    public func activate() -> Bool {
        let (changed, subscribers) = state.withLock { state -> (Set<String>, [UpdateSubscriber]) in
            guard let next = state.pending else { return ([], []) }
            let previous = state.active?.values ?? [:]
            let changed = Set(previous.keys).union(next.values.keys).filter { previous[$0] != next.values[$0] }
            state.pending = nil
            state.active = next
            return (changed, Array(state.subscribers.values))
        }
        // Never invoke stream termination handlers while holding the state lock.
        if !changed.isEmpty { subscribers.forEach { $0.send(changed) } }
        return !changed.isEmpty
    }

    @discardableResult
    public func fetchAndActivate(minimumFetchInterval: TimeInterval? = nil) async throws -> Bool {
        guard try await fetch(minimumFetchInterval: minimumFetchInterval) != .discarded else { return false }
        try Task.checkCancellation()
        return activate()
    }

    /// Clears memory and disk; results of requests already in flight are discarded.
    public func reset() async throws {
        try await background { [self] in
            try cacheAccess.withLock {
                try state.withLock { state in
                    try Task.checkCancellation()
                    state.generation &+= 1
                    state.active = nil
                    state.pending = nil
                    state.lastFetch = nil
                }
                store.clear()
            }
        }
    }

    public func getString(_ key: String) -> String { value(key, parse: { $0 }, zero: "") }
    public func getBoolean(_ key: String) -> Bool { value(key, parse: Self.boolean, zero: false) }
    public func getLong(_ key: String) -> Int64 { value(key, parse: { Int64(Self.trim($0)) }, zero: 0) }
    public func getDouble(_ key: String) -> Double { value(key, parse: Self.double, zero: 0) }

    /// Raw entry origin; typed getters can fall back when the remote value is malformed.
    public func getSource(_ key: String) -> NonaValueSource {
        state.withLock { entry(key, in: $0)?.1 ?? .static }
    }

    public func resolveString(_ key: String) -> NonaResolution<String> { resolve(key, parse: { $0 }) }
    public func resolveBoolean(_ key: String) -> NonaResolution<Bool> { resolve(key, parse: Self.boolean) }
    public func resolveLong(_ key: String) -> NonaResolution<Int64> { resolve(key, parse: { Int64(Self.trim($0)) }) }
    public func resolveDouble(_ key: String) -> NonaResolution<Double> { resolve(key, parse: Self.double) }

    private func value<T>(_ key: String, parse: (String) -> T?, zero: T) -> T {
        state.withLock { state in
            if let entry = state.active?.values[key], let value = parse(entry.value) { return value }
            if let entry = state.defaults[key], let value = parse(entry.value) { return value }
            return zero
        }
    }

    private func resolve<T: Sendable>(_ key: String, parse: (String) -> T?) -> NonaResolution<T> {
        state.withLock { state in
            guard let (entry, source) = entry(key, in: state) else {
                return .failure(reason: state.active == nil ? .notReady : .notFound,
                                message: "No readable value for '\(key)'. Client snapshots contain only frontend-scoped entries.")
            }
            guard let value = parse(entry.value) else {
                return .failure(reason: .typeMismatch, message: "Nona value '\(key)' has the wrong type.")
            }
            return .success(value: value, source: source, contentType: entry.contentType)
        }
    }

    private func entry(_ key: String, in state: State) -> (NonaEntry, NonaValueSource)? {
        if let entry = state.active?.values[key] { return (entry, .remote) }
        if let entry = state.defaults[key] { return (entry, .default) }
        return nil
    }

    private static func trim(_ value: String) -> String { value.trimmingCharacters(in: .whitespacesAndNewlines) }
    private static func boolean(_ value: String) -> Bool? {
        switch trim(value).lowercased() { case "true": return true; case "false": return false; default: return nil }
    }
    private static func double(_ value: String) -> Double? {
        let text = trim(value)
        // Match the decimal grammar used by the Kotlin SDK, excluding hex and suffixes.
        guard text.range(of: #"^[+-]?(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)(?:[eE][+-]?[0-9]+)?$"#,
                         options: .regularExpression) != nil,
              let result = Double(text), result.isFinite else { return nil }
        return result
    }

    private func load(minimumFetchInterval: TimeInterval) async throws -> NonaFetchStatus {
        let now = clock()
        let (generation, previous, lastFetch) = state.withLock { ($0.generation, $0.pending ?? $0.active, $0.lastFetch) }
        if let lastFetch, now >= lastFetch, now - lastFetch < minimumFetchInterval { return .throttled }
        var headers: [String: String] = [:]
        if let apiKey = options.apiKey { headers["X-Api-Key"] = apiKey }
        if let etag = previous?.etag { headers["If-None-Match"] = etag }
        let response = try await http.get(url: options.snapshotURL, headers: headers)
        try Task.checkCancellation()
        let values: [String: NonaEntry]
        let etag: String?
        if response.statusCode == 304 {
            guard let previous else { throw NonaError.unexpectedNotModified }
            values = previous.values
            etag = previous.etag
        } else {
            guard (200...299).contains(response.statusCode) else {
                let problem = try? JSONDecoder().decode(ProblemDetails.self, from: response.body)
                throw NonaError.http(statusCode: response.statusCode,
                                     errorCode: problem?.errorCode,
                                     detail: problem?.detail)
            }
            guard response.body.count <= options.maxResponseBytes else {
                throw NonaError.responseTooLarge(maxBytes: options.maxResponseBytes)
            }
            do { values = try JSONDecoder().decode([String: NonaEntry].self, from: response.body) }
            catch { throw NonaError.invalidSnapshot }
            etag = response.etag
        }
        let snapshot = Snapshot(identity: identity, values: values, etag: etag, fetchedAt: now)
        return try await background { [self] in
            try cacheAccess.withLock {
                let data = try? JSONEncoder().encode(snapshot)
                let committed = try state.withLock { state -> Bool in
                    try Task.checkCancellation()
                    guard state.generation == generation else { return false }
                    if response.statusCode == 304, state.pending == nil { state.active = snapshot }
                    else { state.pending = snapshot }
                    state.lastFetch = now
                    return true
                }
                guard committed else { return .discarded }
                // Keep writes ordered with reset/initialize without blocking value reads.
                if let data { store.write(data) }
                return response.statusCode == 304 ? .notModified : .success
            }
        }
    }
}

private struct ProblemDetails: Decodable {
    let errorCode: String?
    let detail: String?
}

/// Disk operations must not inherit the UI actor. Cancellation propagates to the worker.
private func background<T: Sendable>(_ operation: @escaping @Sendable () throws -> T) async throws -> T {
    let worker = Task.detached(priority: Task.currentPriority, operation: operation)
    return try await withTaskCancellationHandler(operation: { try await worker.value }, onCancel: { worker.cancel() })
}

/// FIFO fetch serialization; cancelled waiters do not start network requests.
private actor FetchGate {
    private var busy = false
    private var waiters: [(UUID, CheckedContinuation<Void, Error>)] = []

    func acquire() async throws {
        try Task.checkCancellation()
        if !busy { busy = true; return }
        let id = UUID()
        try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
                if Task.isCancelled { continuation.resume(throwing: CancellationError()) }
                else { waiters.append((id, continuation)) }
            }
        } onCancel: {
            Task { await self.cancel(id) }
        }
    }

    func release() {
        if waiters.isEmpty { busy = false }
        else { waiters.removeFirst().1.resume() }
    }

    private func cancel(_ id: UUID) {
        guard let index = waiters.firstIndex(where: { $0.0 == id }) else { return }
        waiters.remove(at: index).1.resume(throwing: CancellationError())
    }
}
